walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,115 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.ArrayList;
+import java.util.Arrays;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (request.getDuration() <= TimeRange.WHOLE_DAY.duration()){
+        Collection<TimeRange> mandatoryTimes = findAvailableMeetingTimes(events, request, true);
+        Collection<TimeRange> optionalTimes =  findAvailableMeetingTimes(events,request, false);
+        Collection<String> mandatoryAttendees = request.getAttendees();
+        Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+        if(optionalAttendees.size()== 0){
+            return mandatoryTimes;
+        }
+        if(mandatoryAttendees.size()== 0){
+            return optionalTimes;
+        }
+        Collection<TimeRange> commonTimes = findCommonTimes(mandatoryTimes,optionalTimes, request);
+        if(commonTimes.size() == 0){
+            return mandatoryTimes;
+        }
+        else{
+            return commonTimes;
+        }
+    
+    }
+    else {
+      return Collections.emptyList();
+    }
   }
-}
+  
+  //for mandatoryOrOptional, true means mandatory attendees and false means optional attendees
+   public Collection<TimeRange> findAvailableMeetingTimes(Collection<Event> events, MeetingRequest request, boolean mandatoryOrOptional){
+      Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>(Arrays.asList(TimeRange.WHOLE_DAY));
+      Collection<String> meetingAttendees;
+      if (mandatoryOrOptional == true){
+          meetingAttendees = request.getAttendees();
+      }
+      else{
+          meetingAttendees = request.getOptionalAttendees();
+      }
+      for (Event event : events){
+        Set<String> eventAttendees = event.getAttendees();
+        if (Collections.disjoint(meetingAttendees, eventAttendees) ==  false){ 
+        ArrayList<TimeRange> toRemove = new ArrayList<TimeRange>();
+          ArrayList<TimeRange> toAdd= new ArrayList<TimeRange>();
+          for(TimeRange freeslot : meetingTimes){
+            if (freeslot.overlaps(event.getWhen())){
+              if (event.getWhen().start()<=freeslot.end() && event.getWhen().start()>=freeslot.start()){
+                if (freeslot.start() != event.getWhen().start()){
```

Comment:
Why can't they start at the same time? Also, can't you just remove the `=` form the same comparison one line above?

Reply:
If they start at the same time, it would add a slot such as [30,30) which shouldn't be added. But yes, I will change the line above